packages/vesting-base/src/contract.rs

```diff
@@ -0,0 +1,535 @@
+        order_by: Option<OrderBy>,
+    ) -> StdResult<VestingAccountsResponse> {
+        let start_after = addr_opt_validate(deps.api, &start_after)?;
+
+        let vesting_infos = self.read_vesting_infos(deps, start_after, limit, order_by)?;
+
+        let vesting_accounts: Vec<_> = vesting_infos
+            .into_iter()
+            .map(|(address, info)| VestingAccountResponse { address, info })
+            .collect();
+
+        Ok(VestingAccountsResponse { vesting_accounts })
+    }
+
+    /// Returns the available amount of vested and yet to be claimed tokens for a specific vesting recipient.
+    ///
+    /// * **address** vesting recipient for which to return the available amount of tokens to claim.
+    pub fn query_vesting_available_amount(
+        &self,
+        deps: Deps,
+        env: Env,
+        address: String,
+    ) -> StdResult<Uint128> {
+        let address = deps.api.addr_validate(&address)?;
+
+        let info = self.vesting_info.load(deps.storage, &address)?;
+        let available_amount = compute_available_amount(env.block.time.seconds(), &info)?;
+        Ok(available_amount)
+    }
+
+    /// Manages contract migration.
+    pub fn migrate(
+        &self,
+        _deps: DepsMut,
+        _env: Env,
+        _msg: MigrateMsg,
+    ) -> Result<Response, ContractError> {
+        Ok(Response::default())
+    }
+}
+
+/// Asserts the validity of a list of vesting schedules.
+///
+/// * **addr** receiver of the vested tokens.
+///
+/// * **vesting_schedules** vesting schedules to validate.
+fn assert_vesting_schedules(
+    addr: &Addr,
+    vesting_schedules: &[VestingSchedule],
+) -> Result<(), ContractError> {
+    for sch in vesting_schedules {
+        if let Some(end_point) = &sch.end_point {
+            if !(sch.start_point.time < end_point.time && sch.start_point.amount < end_point.amount)
+            {
+                return Err(ContractError::VestingScheduleError(addr.to_string()));
+            }
+        }
+    }
+
+    Ok(())
+}
+
+/// Computes the amount of vested and yet unclaimed tokens for a specific vesting recipient.
+/// Returns the computed amount if the operation is successful.
+///
+/// * **current_time** timestamp from which to start querying for vesting schedules.
+/// Schedules that started later than current_time will be omitted.
+///
+/// * **vesting_info** vesting schedules for which to compute the amount of tokens
+/// that are vested and can be claimed by the recipient.
+fn compute_available_amount(current_time: u64, vesting_info: &VestingInfo) -> StdResult<Uint128> {
+    let mut available_amount: Uint128 = Uint128::zero();
+    for sch in &vesting_info.schedules {
+        if sch.start_point.time > current_time {
+            continue;
+        }
+
+        available_amount = available_amount.checked_add(sch.start_point.amount)?;
+
+        if let Some(end_point) = &sch.end_point {
+            let passed_time = current_time.min(end_point.time) - sch.start_point.time;
```

Comment:
I am concerned about `end_point.time` potentially being smaller than `sch.start_point.time`, I don't see any code guarding this condition.

Reply:
there is a guardian: https://github.com/neutron-org/neutron-tge-contracts/blob/3f44a88731fc908c235a86427d2345bf08d484e8/packages/vesting-base/src/contract.rs#L247